### PR TITLE
fix: handle empty type_args in define-trait definition

### PR DESCRIPTION
**Pull request by hugocaillard**
Safety check in parse_trait_type_repr

### CODE REVIEW

**Review by jcnelson — COMMENTED** [review]
I think this belongs in Clarinet?

File: clarity/src/vm/types/signatures.rs [hunk 1]
```diff
@@ -1655,6 +1655,9 @@ impl TypeSignature {
         epoch: StacksEpochId,
         clarity_version: ClarityVersion,
     ) -> Result<BTreeMap<ClarityName, FunctionSignature>> {
+        if type_args.is_empty() {
+            return Err(CheckErrors::InvalidTypeDescription);
+        }
         let mut trait_signature: BTreeMap<ClarityName, FunctionSignature> = BTreeMap::new();
         let functions_types = type_args[0]
```
Comment (kantai) [inline clarity/src/vm/types/signatures.rs:1662]: This looks good to me, but I'd suggest removing the indexing operation instead:

```suggestion
        let mut trait_signature: BTreeMap<ClarityName, FunctionSignature> = BTreeMap::new();
        let functions_types = type_args
               .get(0)
               .ok_or_else(|| CheckErrors::InvalidTypeDescription)?       
```

**Comment by hugocaillard** [comment]
@jcnelson I think both are possible. We could probably add a check in clarinet, but on there other, I think it's fine to have this code path being safe as well

**Comment by jcnelson** [comment]
@hugocaillard The runtime crash that this patch addresses does not manifest in the Stacks blockchain.  It _only_ exists in Clarinet.  Therefore, the fix for it should be in Clarinet.

**Comment by kantai** [comment]
> @hugocaillard The runtime crash that this patch addresses does not manifest in the Stacks blockchain. It only exists in Clarinet. Therefore, the fix for it should be in Clarinet.

I don't agree with the conclusion -- while its true that this check is _currently_ unnecessary in this repo, the function has an implicit pre-condition that the caller has performed this check. Ultimately, having this hanging around in the codebase is dangerous, because a future use of this function could forget to perform the pre-check. It's also perfectly safe to add this check to codebase right now for exactly the reasoning you mention: the stacks-node already performs the check before calling this function, so the check failure is unreachable from there anyways.

**Comment by jcnelson** [comment]
The issue in Clarity's trait definition parser pertaining to this crash have less to do with the fact that it's performing a slice access, and more to do with the fact that the safety of the code path in the Stacks blockchain depends on the input first being validated by a call to `DefineFunctionsParsed::try_parse()`.  At a minimum, the problem in the code here is that this pre-condition is not documented in `parse_trait_type_repr`.  However, my ideal outcome is that parser code like `DefineFunctionsParsed::try_parse()` simply returns the first argument (e.g. function name) separately from its arguments, thereby obviating the need for a direct slice in code that consumes this output.  This isn't specific to the trait parser; the whole of the Clarity codebase would benefit from this structural change.

**Comment by kantai** [comment]
> However, my ideal outcome is that parser code like DefineFunctionsParsed::try_parse() simply returns the first argument (e.g. function name) separately from its arguments, thereby obviating the need for a direct slice in code that consumes this output. This isn't specific to the trait parser; the whole of the Clarity codebase would benefit from this structural change.

This would be a larger refactor and require much more testing. I agree that it would be a worthwhile refactor, but I don't want to delay the easy removal of a trap with the hope of a later larger refactoring.

**Comment by jcnelson** [comment]
@kantai I'm not opposed to patching the Stacks blockchain, for the reasons you point out (I posted my above comment before I saw yours).

What I'm opposed to is merging this _before_ Clarinet is patched, or _instead_ of patching Clarinet.

**Comment by obycode** [comment]
> But I don't like the idea of Clarinet passing the buck to us to fix problems that they have the responsibility to address.

I'd argue that this was a case of something potentially dangerous in the stacks-core code, that happened to be caught by Clarinet. This usage could be fixed in Clarinet, but if we want to have the fix in stacks-core any way, then there seems to be no need to also change the way this function is used in Clarinet.

**Review by obycode — APPROVED** [review]

**Review by kantai — APPROVED** [review]

**Pull request merged**